plutus-contract/src/Language/Plutus/Contract/StateMachine.hs

```diff
@@ -0,0 +1,206 @@
+    StateMachineClient(..)
+    , ValueAllocation(..)
+    , SMContractError(..)
+    , AsSMContractError(..)
+    , SM.StateMachine(..)
+    , SM.StateMachineInstance(..)
+    -- * Constructing the machine instance
+    , SM.mkValidator
+    -- * Constructing the state machine client
+    , mkStateMachineClient
+    , defaultChooser
+    -- * Running the state machine
+    , runStep
+    , runInitialise
+    ) where
+
+import           Control.Lens
+import           Control.Monad                     (unless)
+import           Control.Monad.Error.Lens
+import           Data.Text                         (Text)
+import qualified Data.Text                         as Text
+
+import           Language.Plutus.Contract
+import qualified Language.Plutus.Contract.Tx       as Tx
+import qualified Language.Plutus.Contract.Typed.Tx as Tx
+import qualified Language.PlutusTx                 as PlutusTx
+import           Language.PlutusTx.StateMachine    (StateMachine (..), StateMachineInstance (..))
+import qualified Language.PlutusTx.StateMachine    as SM
+import           Ledger                            (Value)
+import           Ledger.Typed.Tx                   (TypedScriptTxOut (..))
+import qualified Ledger.Typed.Tx                   as Typed
+import qualified Ledger.Value                      as Value
+
+import qualified Wallet.Typed.StateMachine         as SM
+
+-- $statemachine
+-- To write your contract as a state machine you need
+-- * Two types @state@ and @input@ for the state and inputs of the machine
+-- * A 'SM.StateMachineInstance state input' describing the transitions and
+--   checks of the state machine (this is the on-chain code)
+-- * A 'StateMachineClient state input' with the state machine instance and
+--   an allocation function
+--
+-- You can then use 'runInitialise' and 'runStep' to initialise and transition
+-- the state machine. 'runStep' gets the current state from the utxo set and
+-- makes the transition to the next state using the given input and taking care
+-- of all payments.
+
+data SMContractError s i =
+    SMError (SM.SMError s i)
+    | InvalidTransition
+    | NonZeroValueAllocatedInFinalState
+    | ChooserError Text
+    deriving (Show)
+
+makeClassyPrisms ''SMContractError
+
+-- | Allocation of funds to outputs of the state machine transition transaction.
+data ValueAllocation =
+    ValueAllocation
+        { vaOwnAddress    :: Value
+        -- ^ How much should stay locked in the contract
+        , vaOtherPayments :: UnbalancedTx
+        -- ^ Any other payments
+        }
+
+instance Semigroup ValueAllocation where
+    l <> r =
+        ValueAllocation
+            { vaOwnAddress = vaOwnAddress l <> vaOwnAddress r
+            , vaOtherPayments = vaOtherPayments l <> vaOtherPayments r
+            }
+
+instance Monoid ValueAllocation where
+    mappend = (<>)
+    mempty  = ValueAllocation mempty mempty
+
+-- | Client-side definition of a state machine.
+data StateMachineClient s i = StateMachineClient
```

Comment:
I'm unsure about the name of this thing. Perhaps it should be `StateMachineContract`?

Reply:
I like `StateMachineClient` better, the contract is what you get from runInitialise/runStep